src/main/java/graphql/execution/AsynchronousExecutionStrategy.java

```diff
@@ -0,0 +1,223 @@
+package graphql.execution;
+
+import graphql.ExecutionResult;
+import graphql.ExecutionResultImpl;
+import graphql.execution.instrumentation.Instrumentation;
+import graphql.execution.instrumentation.InstrumentationContext;
+import graphql.execution.instrumentation.parameters.FieldFetchParameters;
+import graphql.execution.instrumentation.parameters.FieldParameters;
+import graphql.language.Field;
+import graphql.schema.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+
+import static graphql.execution.FieldCollectorParameters.newParameters;
+import static graphql.execution.TypeInfo.newTypeInfo;
+
+public class AsynchronousExecutionStrategy extends ExecutionStrategy {
```

Comment:
This comment is an aside to this PR btw.  

@andimarek 

There is a lot of repeated code in here from the base ExecutionStrategy.  Perhaps we should make a ExecutionStrategyHelper and use composition instead of inheritance to allow better reuse??

Hard problem I know